File: L14GoldmanSachsMedium/MinimumWindowSubstring.cpp
// #include<bits/stdc++.h>
// #include <iostream>
// #include <unordered_map>
// #include <string>
// using namespace std;

// string minWindow(string s, string t)
// {
//     if (s.empty() || t.empty() || s.size() < t.size())
//     {
//         return "";
//     }

//     unordered_map<char, int> tMap;
//     for (char c : t)
//     {
//         tMap[c]++;
//     }

//     int left = 0, right = 0, requiredChars = tMap.size();
//     unordered_map<char, int> windowMap;
//     int minLen = INT_MAX;
//     int minLeft = 0;

//     while (right < s.size())
//     {
//         char rightChar = s[right];
//         windowMap[rightChar]++;

//         while (windowMap.size() == requiredChars)
//         {
//             if (right - left + 1 < minLen)
//             {
//                 minLen = right - left + 1;
//                 minLeft = left;
//             }

//             char leftChar = s[left];
//             windowMap[leftChar]--;
//             if (windowMap[leftChar] == 0)
//             {
//                 windowMap.erase(leftChar);
//             }
//             left++;
//         }

//         right++;
//     }

//     return minLen == INT_MAX ? "" : s.substr(minLeft, minLen);
// }

// int main()
// {
//     string s = "ADOBECODEBANC";
//     string t = "ABC";
//     cout << "Minimum Window Substring: " << minWindow(s, t) << endl; // Output: "BANC"
//     return 0;
// }

#include <iostream>
#include <string>
#include <unordered_map>
#include <climits>

using namespace std;

string minWindow(string s, string t)
{
    if (s.empty() || t.empty() || s.length() < t.length())
    {
        return "";
    }

    // Frequency map for characters in t
    unordered_map<char, int> tMap;
    for (char c : t)
    {
        tMap[c]++;
    }

    // Sliding window
    int left = 0, right = 0;
    int requiredChars = tMap.size(); // Number of unique characters in t
    int formed = 0;                  // Number of unique characters formed in the current window
    unordered_map<char, int> windowMap;
    int minLen = INT_MAX;
    int minLeft = 0;

    while (right < s.length())
    {
        char rightChar = s[right];
        windowMap[rightChar]++;

        // If the frequency of the current character in the window matches the frequency in t, increment formed
        if (tMap.count(rightChar) && windowMap[rightChar] == tMap[rightChar])
        {
            formed++;
        }

        // If the current window has all the required characters, try to shrink it
        while (left <= right && formed == requiredChars)
        {
            // Update the minimum window
            if (right - left + 1 < minLen)
            {
                minLen = right - left + 1;
                minLeft = left;
            }

            // Shrink the window
            char leftChar = s[left];
            windowMap[leftChar]--;
            if (tMap.count(leftChar) && windowMap[leftChar] < tMap[leftChar])
            {
                formed--;
            }
            left++;
        }

        right++;
    }

    return minLen == INT_MAX ? "" : s.substr(minLeft, minLen);
}

int main()
{
    string s = "ADOBECODEBANC";
    string t = "ABC";
    cout << "Minimum Window Substring: " << minWindow(s, t) << endl; // Expected Output: "BANC"
    return 0;
}

File: L14GoldmanSachsMedium/MinimumWindowSubstring.java

// import java.util.HashMap;

// public class MinimumWindowSubstring {

//     public static String minWindow(String s, String t) {
//         if (s == null || t == null || s.length() < t.length()) {
//             return "";
//         }

//         // Frequency map for characters in t
//         HashMap<Character, Integer> tMap = new HashMap<>();
//         for (char c : t.toCharArray()) {
//             tMap.put(c, tMap.getOrDefault(c, 0) + 1);
//         }

//         // Sliding window
//         int left = 0, right = 0, requiredChars = tMap.size();
//         HashMap<Character, Integer> windowMap = new HashMap<>();
//         int minLen = Integer.MAX_VALUE;
//         int minLeft = 0;

//         while (right < s.length()) {
//             char rightChar = s.charAt(right);
//             windowMap.put(rightChar, windowMap.getOrDefault(rightChar, 0) + 1);

//             // If the current window has all the required characters, try to shrink it
//             while (windowMap.size() == requiredChars) {
//                 if (right - left + 1 < minLen) {
//                     minLen = right - left + 1;
//                     minLeft = left;
//                 }

//                 // Shrink the window
//                 char leftChar = s.charAt(left);
//                 windowMap.put(leftChar, windowMap.get(leftChar) - 1);
//                 if (windowMap.get(leftChar) == 0) {
//                     windowMap.remove(leftChar);
//                 }
//                 left++;
//             }

//             right++;
//         }

//         return minLen == Integer.MAX_VALUE ? "" : s.substring(minLeft, minLeft + minLen);
//     }

//     public static void main(String[] args) {
//         String s = "ADOBECODEBANC";
//         String t = "ABC";
//         System.out.println("Minimum Window Substring: " + minWindow(s, t)); // Output: "BANC"
//     }
// }







import java.util.HashMap;

public class MinimumWindowSubstring {

    public static String minWindow(String s, String t) {
        if (s == null || t == null || s.length() < t.length()) {
            return "";
        }

        // Frequency map for characters in t
        HashMap<Character, Integer> tMap = new HashMap<>();
        for (char c : t.toCharArray()) {
            tMap.put(c, tMap.getOrDefault(c, 0) + 1);
        }

        // Sliding window
        int left = 0, right = 0, requiredChars = tMap.size();
        int formed = 0; // Number of unique characters formed in the current window
        HashMap<Character, Integer> windowMap = new HashMap<>();
        int minLen = Integer.MAX_VALUE;
        int minLeft = 0;

        while (right < s.length()) {
            char rightChar = s.charAt(right);
            windowMap.put(rightChar, windowMap.getOrDefault(rightChar, 0) + 1);

            // If the frequency of the current character in the window matches the frequency in t, increment formed
            if (tMap.containsKey(rightChar) && windowMap.get(rightChar).intValue() == tMap.get(rightChar).intValue()) {
                formed++;
            }

            // If the current window has all the required characters, try to shrink it
            while (left <= right && formed == requiredChars) {
                if (right - left + 1 < minLen) {
                    minLen = right - left + 1;
                    minLeft = left;
                }

                // Shrink the window
                char leftChar = s.charAt(left);
                windowMap.put(leftChar, windowMap.get(leftChar) - 1);
                if (tMap.containsKey(leftChar) && windowMap.get(leftChar) < tMap.get(leftChar)) {
                    formed--;
                }
                left++;
            }

            right++;
        }

        return minLen == Integer.MAX_VALUE ? "" : s.substring(minLeft, minLeft + minLen);
    }

    public static void main(String[] args) {
        String s = "ADOBECODEBANC";
        String t = "ABC";
        System.out.println("Minimum Window Substring: " + minWindow(s, t)); // Expected Output: "BANC"
    }
}
